src/Common/tests/gtest_concurrency_control.cpp

```diff
@@ -212,6 +212,208 @@ TEST(ConcurrencyControl, SetSlotCount)
+        ASSERT_TRUE(slot1);
+        slots.emplace_back(std::move(slot1));
+        ASSERT_TRUE(!allocations[i]->tryAcquire());
+    }
+}
+
+TEST(ConcurrencyControl, FairRoundRobinReleaseUnacquiredSlots)
+{
+    ConcurrencyControlTest t(10);
+    t.cc.setScheduler("fair_round_robin");
+    {
+        std::vector<SlotAllocationPtr> allocations;
+        allocations.reserve(10);
+        for (int i = 0; i < 10; i++)
+            allocations.emplace_back(t.cc.allocate(1, 2));
+        // Do not acquire - just destroy allocations with granted slots
+    }
+    // Check that slots were actually released
+    auto allocation = t.cc.allocate(0, 20);
+    std::vector<AcquiredSlotPtr> acquired;
+    while (auto slot = allocation->tryAcquire())
+        acquired.emplace_back(std::move(slot));
+    ASSERT_TRUE(acquired.size() == 10);
+}
+
+
+TEST(ConcurrencyControl, FairRoundRobinDestroyNotFullyAllocatedAllocation)
+{
+    ConcurrencyControlTest t(10);
+    t.cc.setScheduler("fair_round_robin");
+    for (int i = 0; i < 3; i++)
+    {
+        auto allocation = t.cc.allocate(5, 20);
+        std::vector<AcquiredSlotPtr> acquired;
+        while (auto slot = allocation->tryAcquire())
+            acquired.emplace_back(std::move(slot));
+        ASSERT_TRUE(acquired.size() == 15);
+    }
+}
+
+TEST(ConcurrencyControl, FairRoundRobinDestroyAllocationBeforeSlots)
+{
+    ConcurrencyControlTest t(10);
+    t.cc.setScheduler("fair_round_robin");
+    for (int i = 0; i < 3; i++)
+    {
+        std::vector<AcquiredSlotPtr> acquired;
+        auto allocation = t.cc.allocate(5, 20);
+        while (auto slot = allocation->tryAcquire())
+            acquired.emplace_back(std::move(slot));
+        ASSERT_TRUE(acquired.size() == 15);
+        allocation.reset(); // slots are still acquired (they should actually hold allocation)
+    }
+}
+
+TEST(ConcurrencyControl, FairRoundRobinGrantReleasedToTheSameAllocation)
+{
+    ConcurrencyControlTest t(3);
+    t.cc.setScheduler("fair_round_robin");
+    auto allocation = t.cc.allocate(1, 11);
+    std::list<AcquiredSlotPtr> acquired;
+    while (auto slot = allocation->tryAcquire())
+        acquired.emplace_back(std::move(slot));
+    ASSERT_TRUE(acquired.size() == 4); // X 0 1 2
+    acquired.clear();
+    while (auto slot = allocation->tryAcquire())
+        acquired.emplace_back(std::move(slot));
+    ASSERT_TRUE(acquired.size() == 3); // 3 4 5
+    acquired.pop_back();
+    while (auto slot = allocation->tryAcquire())
+        acquired.emplace_back(std::move(slot));
+    ASSERT_TRUE(acquired.size() == 3); // 3 4 6
+    acquired.pop_front();
+    while (auto slot = allocation->tryAcquire())
+        acquired.emplace_back(std::move(slot));
+    ASSERT_TRUE(acquired.size() == 3); // 4 6 7
+    acquired.clear();
+    while (auto slot = allocation->tryAcquire())
+        acquired.emplace_back(std::move(slot));
+    ASSERT_TRUE(acquired.size() == 2); // 8 9
+}
+
+TEST(ConcurrencyControl, FairRoundRobinFairGranting)
```

Comment:
As we wanted to solve the issue of some operations blocking others because of the min threads, maybe it would make sense to have `1` as min. What do you think?

Reply:
Yes, make sense, lets improve it.